### DISCUSSION TITLE
Schedule drop of PHP 5

### DISCUSSION THREAD

**Discussion by Alkarex** [FreshRSS/FreshRSS#3321] [Strategy]
Not yet for the upcoming release 1.18.0, but some months later, it would be time to revise our minimum requirements for PHP.
Our last discussion on the topic was back in September 2019: https://github.com/FreshRSS/FreshRSS/pull/2495 + https://github.com/FreshRSS/FreshRSS/pull/2527 . Since then, Debian 8 Jessie (shipping PHP 5.6.7) has ended, PHP 8 has been released, and W3Techs stats indicate that PHP7 is now more widely used than PHP5 at last https://w3techs.com/technologies/history_details/pl-php/all/q (only by 20 %-points though, as of January 2021).

I personally look forward to using a few new syntax features such as `??` to simplify our many `isset() ? ... : ...` constructs, and better [type declaration](https://php.net/language.types.declarations) to improve the code: https://php.net/migration70.new-features
See e.g. https://github.com/FreshRSS/FreshRSS/pull/3315/files#r550897643

Distributions (using e.g. https://distrowatch.com/table.php?distribution=debian ):
* no impact on Debian, Ubuntu, Fedora, Alpine, OpenWRT, FreeBSD, OpenSuze, Mageia, as all active versions have PHP ≥ 7
* we had already dropped Red Hat 7, but we keep supporting RHEL 8 (released 2019-05)
* we had already dropped CentOS 7, but we keep supporting CentOS 8 (released 2020-06)
* Synology has been shipping PHP 7+ in addition of PHP 5.6 for several years

As of September 2020, WordPress sticks to PHP 5.6+ https://make.wordpress.org/core/2020/08/24/proposal-dropping-support-for-old-php-versions-via-a-fixed-schedule/#comment-39736 (not the article itself, but the last comment is the one that counts)

As of January 2021, dropping PHP 5 (compared to our current PHP 5.6+) would "cost" about 21% of the current PHP servers according to https://w3techs.com/technologies/details/pl-php/5 (fraction of PHP 5.6 in PHP 5.x times fraction of PHP 5.x in PHP), which is still quite a lot but dropping fast.

On a personal note, I give a high priority to Debian (and Ubuntu LTS) https://wiki.debian.org/LTS , https://packages.debian.org/search?keywords=php&exact=1 which would also give an idea about when to consider dropping later PHP versions (e.g. drop PHP 7.0- after June 2022)

Let's keep this discussion open to track the evolution...

**Comment by aledeg** [comment]
Here is the list of things I am looking forward to use:
- `<=>` a.k.a. spaceship operator,
- `?:` a.k.a. elvis operator,
- `??` a.k.a. null coalesce operator,
- type declaration.

I can't wait! :)

**Comment by Alkarex** [comment]
I have just checked again different metrics. PHP 5.6 usage has dropped by 1%-point since last month, and since there is not much new on WordPress' side (which I would otherwise have liked to follow), I suggest we schedule dropping PHP5 for FreshRSS 1.19.0 (so let's wait a bit more after the release of 1.18.0 in the likely case we will have minor 1.18.x versions), and then move to 7.0 as the oldest supported version :-)

**Reply by aledeg** [reply]
Sounds reasonable.

**Comment by Alkarex** [comment]
I suggest we drop IE11 at the same time https://slashdot.org/story/21/03/13/033223/wordpress-considers-dropping-support-for-internet-explorer-11?utm_source=rss1.0mainlinkanon&utm_medium=feed (i.e. for FreshRSS 1.19.0)

**Comment by Alkarex** [comment]
Release done https://github.com/FreshRSS/FreshRSS/releases/tag/1.18.1 Thanks all!

**Note:** More time was allocated during this release as this will be the last release before dropping support for PHP 5 and Internet Explorer 11 (the end of an era!)

**Reply by Alkarex** [reply]
https://github.com/FreshRSS/FreshRSS/pull/3666

**Comment by Alkarex** [comment]
For the record,
* ~~Debian 9 Stretch supports PHP 7.0 until June 2022~~
* ~~Ubuntu 18.04 LTS supports PHP 7.2 until April 2023~~
* Debian 10 Buster supports PHP 7.3 until June 2024
* Ubuntu 20.04 LTS supports PHP 7.4 until April 2025
* Debian 11 Bullseye supports PHP 7.4 until June 2026
* Ubuntu 22.04 LTS supports PHP 8.1 until April 2027

Let's come back to it in a year's time :-)

**Reply by ColonelMoutarde** [reply]
A small question, as the support of php7.0 stops with debian in June 2022, would it not be more relevant to pass directly to php7.2 in mini version because the addition of "?" in the parameters of function is a not negligible plus

**Reply by Alkarex** [reply]
Indeed, good observation! All the main supported distributions (cf. top post) seem to have PHP 7.2+ by July 2022.
See also https://w3techs.com/technologies/details/pl-php/7

**Reply by Alkarex** [reply]
Let's target 1.21.0 for PHP 7.2+ https://github.com/FreshRSS/FreshRSS/milestone/53

**Reply by math-GH** [reply]
I would suggest
- 1.20 PHP 7.0+(Autumn 2022)
- 1.21 PHP 7.2+ (Spring 2023)
- 1.22 PHP 7.4+ (Autumn 2023)

**Reply by Alkarex** [reply]
Interesting statistics: https://wordpress.org/about/stats/#php_versions
From https://github.com/simplepie/simplepie/pull/743#issuecomment-1231869183

**Reply by Alkarex** [reply]
PHP 7.2+ merged https://github.com/FreshRSS/FreshRSS/pull/4848

**Comment by math-GH** [comment]
how is the progress? What are the concrete next steps? Just to change the documentation that only PHP7+ is supported?

**Comment by Frenzie** [comment]
Pretty much. First say what's (un)supported, then start using some new
features without worrying about it.

**Reply by math-GH** [reply]
How can I enable the log levels to see the notices and warning? Where can I find the logs?

**Reply by Frenzie** [reply]
There are environment variables; not sure otoh if there's any other way.

**Reply by Alkarex** [reply]
`-e FRESHRSS_ENV=development` if using Docker or similar.
Otherwise, see `environment` in `./data/config.php`
